### ISSUE TITLE
System audio gets muted/reduced in Unity due to audio communications protocol.

### ISSUE THREAD

**Issue by rygo6-MS** [microsoft/MixedReality-WebRTC#573]
I've just merged in all the latest upstream changes from master, I have audio successfully going through Unity and coming out of an AudioSource spatially in the environment. However, I still notice that when the WebRTC stream connects my system audio gets reduced or muted, due to it claiming audio via the 'Communications Protocol' in windows.

I thought that once audio was going through Unity, then it would not touch that communication protocol, or anything to do with system audio whatsoever. Is that not true? Or is something wrong on my end? 

We've been running this in Altspace in beta test and have found that users who have a bluetooth or usb headset plugin will have their 'Windows Audio Service' crash when a WebRTC stream connects. I assume that is caused by it affecting the communication protocol, or system audio in some way. Have you heard of any issues similar to that? 

Also, I am working now on getting our code ready to PR back, we have been on a bit of detour past 3 weeks due to a big Altspace event.

**Comment by djee-ms** [comment]
I've just tried `master` with some Surface Headphones (bluetooth), after disabling other devices to make sure it picks it. I called from our Unity sample VideoChat scene (Unity editor x64 Desktop) to our sample TestAppUWP (UWP ARM Release) and have audio and video working as expected.

**Comment by rygo6-MS** [comment]
For the bluetooth crashing issue I have been trying to get more feedback from the users that claimed to have encountered it. Turns out it is actually hard for them to reproduce, so it could be related to something deeper than WebRTC, drivers of their headset or other. Since neither of you haven't heard anything of it, I'll assume it may not be related to this for now, until I get more logs from them.

About the audio muting, I made a mistake and was actually not running the test with the latest from master properly. It no longer lowers the system sound. But I've actually encountered a new issue that is worse. My `'protected void OnAudioFilterRead(float[] data, int channels)'` method in AudioRenderer.cs is getting called, from the appropriate audio source, but the data array is filled with all zeroes. So I am not getting any sound at all. But it is registering the correct RemotAudioTrack and it is calling that method. Just all zeroes in the data array. Do you have any ideas on what that could be caused by? I switch it back to the Release2 dll's and sound comes through without issue.

**Comment by djee-ms** [comment]
@rygo6-MS is that happening in the Unity editor or inside a deployed app? Between 2.0.0 and `master`, the audio-related changes are:
- For Windows Desktop, we now use ADM2 by default, which is a different audio engine (#507). You can fall back to ADM1 like 2.0.0 by calling `Library.UseAudioDeviceModule(AudioDeviceModule.LegacyModule)`.
- For Windows UWP, we now use a custom audio mixer ([74bece9](74bece9a58dfc0e548eb275dd4bc4f525282a256)) to be able to disable the default audio output to system speakers and have only the audio injected into the Unity DSP pipeline. That might have caused a regression.

**Comment by fibann** [comment]
@rygo6-MS 
> is that happening in the Unity editor or inside a deployed app

More generally, what platform (UWP, Unity editor, Win32 app, ...) are you using for sending and what for receiving? Note that there might be an issue with the sending app (some problem with mic capture) as well as with the receiving one.

**Comment by rygo6-MS** [comment]
Calling `Library.UsedAudioDeviceModule = AudioDeviceModule.LegacyModule;` does give me back my sound. 

But I assume using ADM2 would be better? This is a subject I was going to bring up in another ticket, but sound streaming through MRWebRTC would encounter periodic "time dilation", speeding up or slowing down. Does ADM2 help with that?

How would I go about getting ADM2 to work? I am trying this in the Unity Editor and also a build. No sound in either on Win10 x64. I also have this issue in the Altspace codebase, as well as a smaller test Unity application with nothing but MRWebRTC and the code to signal to our backend.

Our stream currently comes from a Chrome browser plugin being routed through Janus VideoRoom on Azure.

**Comment by djee-ms** [comment]
ADM2 is only for Windows Desktop, and is essentially for us a workaround for some annoying behavior of ADM1 which fails its initialization if any unsupported device is even plugged in to the host machine, even if unused. What that means is a user with _e.g._ an Azure Kinect DK (unsupported because of the 7-microphone array) plugged in to their PC will see ADM1 fail at startup and make MixedReality-WebRTC unusable. In terms of functionalities, I don't think there's a large difference between ADM1 and ADM2 though.

About the time dilatation it is with ADM1 right? Unfortunately that doesn't ring a bell, and would be an issue to investigate with Google. But their policy is to not put work into ADM1, so unlikely to get fixed anyway.

We have little to no idea why ADM2 would produce all-zero buffers. The only time that happened to me in the past is when audio negotiation was not done correctly and the audio track was either not sending or more likely was software-muted (there is a mute feature at the audio track level). Another possibility is overly aggressive or misconfigured echo cancellation (AEC) discarding all recorded sound. In all cases unfortunately we don't expose much in terms of tweaking.

For audio muffling, entirely untested code but you could try setting the recording and playback devices in [`GlobalFactory::InitializeImplNoLock()`](https://github.com/microsoft/MixedReality-WebRTC/blob/480981ac74a6040b3058d620c0de2fa617d1f301/libs/mrwebrtc/src/interop/global_factory.cpp#L302) as the default one `kDefaultDevice` (and not the default communication one `kDefaultCommunicationDevice`, which is the default value for both):
```diff
- return webrtc::CreateWindowsCoreAudioAudioDeviceModule();
+ rtc::scoped_refptr<webrtc::AudioDeviceModule> adm2 = webrtc::CreateWindowsCoreAudioAudioDeviceModule();
+ adm2->SetPlayoutDevice(webrtc::AudioDeviceModule::kDefaultDevice);
+ adm2->SetRecordingDevice(webrtc::AudioDeviceModule::kDefaultDevice);
+ return adm2;
```

**Comment by rygo6-MS** [comment]
So I've not been able to figure out what causes ADM2 not to work. I enable LegacyModule, sound works, then enable DefaultModule, sound doesn't work. No errors I can discern in the unity logs. In trying to debug this into C++ there looks to be a lot of log checks on the native `mrsAudioTrackReadBufferRead` method. Is there a way to get the output of those C++ logs while running as a Unity plugin? Or is there any other way I could pull some other logs from a deeper level to debug?

**Comment by djee-ms** [comment]
Those are piped to `OutputDebugString()` in Debug so attaching a debugger should allow you to see the messages, in the Visual Studio Output window for example.

**Comment by fibann** [comment]
> attaching a debugger

Also note that you have to set the debugger to debug _native code_. The default Unity debugger won't show you the native output.

**Comment by djee-ms** [comment]
Yes I meant actually attaching manually a native debugger to the Unity Editor native x64 process (or your application process if you're using a build).

**Comment by rygo6-MS** [comment]
I figured this out. ADM2 does not work for me if I have more than one audio output enabled in Windows. I must disable all the others ones like so:
![Screenshot 2020-09-25 145230](https://user-images.githubusercontent.com/53834195/94319098-c0726880-ff3e-11ea-8a98-74cd445081c3.png)

May be because I have a lot of audio outputs? I have:
- USB DAC
- motherboard audio
- monitor audio
- Valve Index audio
- Oculus Audio
- bigscreen virtualized audio
- voicemod virtualized audio

I may be an atypical audio use case since I have all the VR headsets, all the VR apps which make virtualized audio outputs (Bigscreen and Voicemod) and three separate physical DAC's. But, would be nice to sort this out. I assume it may be something with it not knowing which to select from all the options.

**Comment by djee-ms** [comment]
Does that work with any one source selected? Or only with this one? Otherwise said, is it a problem with the number of sources, or is there a specific one of them causing troubles?

**Comment by rygo6-MS** [comment]
It seems to work with any source. Just only one can be enabled at a time.

**Comment by rygo6-MS** [comment]
I want to bump this again, as we've encountered a small percentage of users now that have the same issue as myself. They go no sound on PC unless they disable all audio outputs except for one. It is looking like I will either have to switch AltSpace back to ADM1 and deal with the crashing, ADM1 would just outright hard crash for a number of users. Or I will have to try to debug this issue with ADM2.

Any ideas on what could be the cause of this in ADM2? 
Or do you know if the crashing issues of ADM1 have been fixed upstream?

**Comment by fibann** [comment]
Sorry, I don't have easy fixes to suggest - ADMs are contained in the WebRTC implementation and I am not incredibly familiar with them, so your guesses are as good as mine here. On top of that, we haven't managed to reproduce the issue on our devices so we can't really debug this. If you can look into the code and check what is going wrong, or find a specific device/configuration that is causing this, that would be a great help.

There has been no work on ADM1 on our end - it's deprecated by Google so we have switched to ADM2 as the recommended way to go. But if you already have some simple crash fix that you can merge we'd be happy to take it.
